velox/expression/tests/ExprOptimizerTest.cpp

```diff
@@ -0,0 +1,396 @@
+  /// @param rowType Row type of `expression`.
+  /// @param lambdaInputTypes Input types for any lambda sub-expressions in
+  /// `expression`.
+  core::TypedExprPtr makeTypedExpr(
+      const std::string& expression,
+      const RowTypePtr& rowType,
+      const std::vector<TypePtr>& lambdaInputTypes) {
+    auto untyped = parse::parseExpr(expression, options_);
+    return core::Expressions::inferTypes(
+        untyped, rowType, lambdaInputTypes, execCtx_->pool());
+  }
+
+  RowVectorPtr fuzzFlat(const RowTypePtr& rowType) {
+    VectorFuzzer::Options options;
+    options.vectorSize = 100;
+    VectorFuzzer fuzzer(options, pool());
+    return fuzzer.fuzzInputFlatRow(rowType);
+  }
+
+  std::shared_ptr<core::QueryCtx> queryCtx_{core::QueryCtx::create()};
+  std::unique_ptr<core::ExecCtx> execCtx_{
+      std::make_unique<core::ExecCtx>(pool_.get(), queryCtx_.get())};
+  parse::ParseOptions options_;
+};
+
+TEST_F(ExprOptimizerTest, constantFolding) {
+  // comparison.
+  testExpression("'a' = substr('abc', 1, 1)", "true");
+  testExpression("'a' = cast(null as varchar)", "cast(null AS BOOLEAN)");
+  testExpression(
+      "cast((12345678901234567890.000 + 0.123) AS DECIMAL(23, 3)) = cast(12345678901234567890.123 AS DECIMAL(23, 3))",
+      "true");
+  testExpression("'0ab01c' LIKE '%ab%c%'", "true");
+  testExpression("3 between (4 - 2) and 2 * 2", "true");
+  testExpression(
+      "3 between cast(null AS BIGINT) and 4", "cast(null AS BOOLEAN)");
+  testExpression("IF(2 = 2, 3, 4)", "3");
+  testExpression("IF(cast(null AS BOOLEAN), 3, 4)", "4");
+  testExpression("IF('hello' = 'foo', cast(null AS BIGINT), 4)", "4");
+  testExpression(
+      "IF('hello' = 'hello', cast(null AS BIGINT), 4)", "cast(null AS BIGINT)");
+
+  // cast.
+  testExpression("cast('-123' as BIGINT)", "-123");
+  testExpression("cast(DECIMAL '1234567890.123' as BIGINT)", "1234567890");
+  testExpression("cast('t' as BOOLEAN)", "true");
+  testExpression("cast('f' as BOOLEAN)", "false");
+  testExpression("cast('1' as BOOLEAN)", "true");
+  testExpression("cast('0' as BOOLEAN)", "false");
+  testExpression("cast(-123.456E0 as VARCHAR)", "'-123.456'");
+  testExpression("cast(cast('abcxyz' as VARCHAR) as VARCHAR)", "'abcxyz'");
+
+  // Partial constant folding in expression tree.
+  testExpression(
+      "1234 between a and 2000 + 1",
+      "1234 between a and 2001",
+      ROW({"a"}, {BIGINT()}));
+  testExpression(
+      "concat(substr(a, 1 + 5 - 2), substr(b, 2 * 3))",
+      "concat(substr(a, 4), substr(b, 6))",
+      ROW({"a", "b"}, {VARCHAR(), VARCHAR()}));
+  testExpression(
+      "a = 'z' and b = (1 + 1) or c = (5 - 1) * 3 or 2 * 3 = 6 / 1 and abs(-4) = 2 * 2",
+      "a = 'z' and b = 2 or c = 12 or true",
+      ROW({"a", "b", "c"}, {VARCHAR(), BIGINT(), INTEGER()}));
+}
+
+TEST_F(ExprOptimizerTest, lambdaExpr) {
+  // basic.
+  auto type = ROW({"a"}, {BIGINT()});
+  std::vector<TypePtr> lambdaTypes = {BIGINT()};
+  testExpression(
```

Comment:
Let's test with lambda functions, not expressions. 

"filter(a, x -> abs(...))"